README.MD
<p align="center">
  <img src="Pictures_for_README/binoculars.png" alt="Ciphey">
</p>


<p align="center">
<a href="https://pypi.org/project/ciphey/"><img src="https://img.shields.io/pypi/v/ciphey.svg"></a>
  <img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="Ciphey">
  <img src="https://github.com/brandonskerritt/Ciphey/workflows/Python%20application/badge.svg?branch=master" alt="Ciphey">
  <img alt="PyPI - Downloads" src="https://img.shields.io/pypi/dm/ciphey">
<img src="https://codecov.io/gh/ciphey/ciphey/branch/master/graph/badge.svg">
<a href="https://ciphey.readthedocs.io/"><img src="https://readthedocs.org/projects/ciphey/badge/"></a>
</p>

# What is this?
Ciphey is an automated decryption tool. Input encrypted text, get the decrypted text back.
> "What type of encryption?"

That's the point. You don't know, you just know it's possibly encrypted. Ciphey will figure it out for you.
Ciphey uses a deep neural network with a simple filtration system to approximate what something is encrypted with. And then a custom-built, customisable natural languge processing Language Checker Interface, which can detect when the given text becomes plaintext.


Ciphey can solve most things in about 2 seconds.
<p align="center" href="https://asciinema.org/a/336257">
  <img src="Pictures_for_README/index.gif" alt="Ciphey demo">
</p>

# Features

- **20+ encryptions supported** such as hashes, encodings (binary, base64) and normal encryptions like Caesar cipher, Transposition and more. **[For the full list, click here](https://ciphey.readthedocs.io/en/latest/ciphers.html)**
- **Deep neural network for targetting the right decryption** resulting in decryptions taking less than 3 seconds. If Ciphey cannot decrypt the text, Ciphey will use the neural network analysis to give you information on how to decrypt it yourself.
- **Custom built natural language processing module** Ciphey can determine whether something is plaintext or not. It has an incredibly high accuracy, along with being fast.
- **Multi Language Support** at present, only English.
- **Supports hashes & encryptions** Which the alternatives such as CyberChef do not. 
- **[C++ core](https://github.com/Ciphey/CipheyCore)** Blazingly fast.


# Getting Started
## Installation
### Pip
```python3 -m pip install -U ciphey```

The -U is needed, as sometimes PyPi gets stuck on an older version.

```ciphey -t "encrypted text here"```
To run ciphey.

### Cloning from GitHub

```
  git clone https://github.com/Ciphey/Ciphey
  cd Ciphey
  python3 -m ciphey -t "encrypted text here"
```
### Running Ciphey
To get rid of the progress bars, probability table, and all the noise use the grep mode.

```ciphey -t "encrypted text here" -g```
For a full list of arguments, run `ciphey -h`.

It is also possible to pipe data into Ciphey, or to use Ciphey like `ciphey 'encrypted text here'`
### Importing Ciphey
You can import Ciphey\'s __main__ and use it in your own programs and code.
This is feature is expected to expand in the next major version.
# Docs
The docs are located at [https://ciphey.readthedocs.io/en/latest/](https://ciphey.readthedocs.io/en/latest/)


# Contributors
Ciphey was invented by [Brandon Skerritt](https://github.com/brandonskerritt) way back in 2008 (don't worry, the code has upgraded since then 😜) but it wouldn't be where it is today without [Cyclic3](https://github.com/Cyclic3).
## Contributing
Please read the contributing file.
## Code Contributors
* [Cyclic3](https://github.com/Cyclic3) for various refactorings and for the C++ core
* [Szymex73](https://github.com/szymex73) for providing the TryHackMe answers dataset.
* [Michal](https://github.com/michalani) for helping with Base85 decoding.
## Financial Contributors
Please donate to us, we're students and we want Huel.


changelog.md
# Change Log
The changelog for Ciphey.

## Coming up next
These are taken from the GitHub Issues tab.
* Improved Neural Network (See /documentation for more information on this)
* Mod + XOR decryptions
* Offline mode (No hash lookups)
* Killswitch - Ciphey kills itself after X seconds. Some decryptions take forever, so if Ciphey is used as a library, waiting 30 - 60 minutes for 1 decryption module to finish sucks. Ciphey will either find the answer in X minutes or seconds, or it will return False.
* Add screenshots to readme.md (ascii cinema)
* Creating a Flask version so Ciphey can be implemented as an API. Useful for creating phone / web apps.
* Ciphey's core (the decryption modules) are now implemented in C++
* The code is now pep8'd
* Move to Poetry from setuptools.py
* Moved to Pytest from unittest
## 4.1
#### Features
* Vigenere is now enabled, due to massive performance gains from the C++ core
* Pytest can now be run over the entire program, from main to the output. This means we can crate tests that test the whole of Ciphey, not just small unit tests.
* Better input handling. Ciphey now supports pipes `echo 'hello' | ciphey` and text with no arguments `ciphey 'hello'.`
#### Bug Fixes
* Chi-squared calcuations are now done _correctly_
* Fixed bug where __main__ didn't return the output.
* Multiple debug statements were printed when not in debug mode.
#### Maintenance
* Offloaded lots of stuff onto C++ core to get much speed
* Disabled addition of language checkers, as no-one could explain why it would make sense
* Bases.py is refactored so users can manually call decryptions. The name has also changed from base64 to bases.
* LanguageChecker now follows PEP8.
* Main and MathsHelper now follow PEP8.
* Now uses Nox to Pytest against multple Python versions (3.6, 3.7, and 3.8).
* Code coverage is now calculated and used.
* Automatic application of Black formatter upon push.
* Automatic uploading to PyPi testing.
## 3.1
#### Features
* Adding a logging library (Loguru) and implemented full logging of the
program when activated. The program is now easier to decrypt.
* Added transposition cipher support
* The program is more verbose, which means in non-grep mode users can see the
program operate in real time.
* Ciphey now takes the first non flagged argument as the text to the input.
* It is now possible to pipe data into ciphey like `echo 'hello' | python3 -m ciphey`
#### Bug Fixes
* Language checker's first phase is now augmented. There was a bug where chi
squared would score low on paragraphs of English text. To amend this, a 2nd
first-phase check was added. If __any__ of the top 1000 English words appear
in the text, go onto the 2nd phase (dictionary checker full).
* Timeouts are now implemented in the hash cracking module.
* Flags previously required arguments when it didn't make sense to have an argument. Now some flags can be ran without arguments.
#### Maintenance
* Updated & added more tests.
* Updated the GitHub action to automatically run tests and Black against the
code base.
* Switch from `alive_bar` to ``Rich``. Rich supports progress bars like alive_bar,
but it also supports rich text output such as tables. The table is used to
prettily display the softmax distribution from the neural network to the user.
The user can now easily read what the neural network thinks the decryption is,
and what's being worked on first.
## 3.0.7
#### Features
* Added other bases
#### Bug Fixes
* Fixed program not installing via Pip
* Fixed basicEncryption module from crashing the program
#### Maintenance
* [Added](Added) program to Pip


noxfile.py
"""
The file for Nox
"""
import nox
from typing import Any
from nox.sessions import Session
import tempfile

locations = "ciphey/", "tests/", "docs/"
nox.options.sessions = "safety", "tests"
package = "ciphey"


def install_with_constraints(session: Session, *args: str, **kwargs: Any) -> None:
    """Install packages constrained by Poetry's lock file.
    This function is a wrapper for nox.sessions.Session.install. It
    invokes pip to install packages inside of the session's virtualenv.
    Additionally, pip is passed a constraints file generated from
    Poetry's lock file, to ensure that the packages are pinned to the
    versions specified in poetry.lock. This allows you to manage the
    packages as Poetry development dependencies.
    Arguments:
        session: The Session object.
        args: Command-line arguments for pip.
        kwargs: Additional keyword arguments for Session.install.
    """
    with tempfile.NamedTemporaryFile() as requirements:
        session.run(
            "poetry",
            "export",
            "--dev",
            "--format=requirements.txt",
            f"--output={requirements.name}",
            external=True,
        )
        session.install(f"--constraint={requirements.name}", *args, **kwargs)


# noxfile.py
@nox.session(python="3.8")
def black(session):
    args = session.posargs or locations
    session.install("black")
    session.run("black", *args)


@nox.session(python="3.8")
def safety(session):
    with tempfile.NamedTemporaryFile() as requirements:
        session.run(
            "poetry",
            "export",
            "--dev",
            "--format=requirements.txt",
            "--without-hashes",
            f"--output={requirements.name}",
            external=True,
        )
        install_with_constraints(session, "safety")
        session.run("safety", "check", f"--file={requirements.name}", "--full-report")


@nox.session(python="3.8")
def coverage(session: Session) -> None:
    """Upload coverage data."""
    install_with_constraints(session, "coverage[toml]", "codecov")
    session.run("pip3", "install", "cipheydists")
    session.run("coverage", "xml", "--fail-under=0")
    session.run("codecov", *session.posargs)


# noxfile.py
@nox.session(python="3.8")
def docs(session: Session) -> None:
    """Build the documentation."""
    install_with_constraints(session, "sphinx")
    session.run("sphinx-build", "docs", "docs/_build")


# python=["3.8", "3.7", "3.6"])
@nox.session(python="3.8")
def tests(session):
    session.run("pip3", "install", "cipheydists")
    session.run("poetry", "install", external=True)
    session.run("poetry", "run", "pytest", "--cov=ciphey")
